Add support for Dynu.com

msxdan opened the pull request:
Added support for dynu. I could only test IPv4 as my ISP doesn't provide me an IPv6 but I added the right param to the query so it should work as good as IPv4 does.

![image](https://user-images.githubusercontent.com/32611659/147878241-3282590b-87e3-4caf-a22a-5723347decf8.png)
![image](https://user-images.githubusercontent.com/32611659/147840471-d051824e-5a3f-4bb2-ba35-cfdff7a41cc0.png)



Closes #222, #237, #263, #269

qdm12 commented on internal/settings/providers/dynu/provider.go:
```diff
@@ -0,0 +1,155 @@
+
+func (p *provider) isValid() error {
+	switch {
+	case len(p.username) == 0:
```
```suggestion
	case p.username == "":
```

qdm12 commented on internal/settings/providers/dynu/provider.go:
```diff
@@ -0,0 +1,155 @@
+	switch {
+	case len(p.username) == 0:
+		return errors.ErrEmptyUsername
+	case len(p.password) == 0:
```
```suggestion
	case p.password == "":
```

qdm12 commented on internal/settings/providers/dynu/provider.go:
```diff
@@ -0,0 +1,155 @@
+	case len(p.password) == 0:
+		return errors.ErrEmptyPassword
+	case p.host == "*":
+		return errors.ErrHostWildcard
```
Wildcards are not supported then right? Are custom hosts (i.e. `a.example.com`) supported?

marvin-noel replied:
If I use @ or * it updates all DDNS IPs for my domains on Dynu, * doesn't update subdomains of a concrete domain, so `*.domain.com` updates domain.com but not `subdomain.domain.com` 

There's an `alias` (subdomain) parameter that could be used to update a concrete subdomain, to update a bunch of domains and subdomains in one call there's an optional parameter called `location` which could be used if configured in Dynu, would be similar to * but you have to set your domains and subdomains to a group in dynu.

I'm adding more features to support all these parameters, I'll update the pull as soon as I finish.

qdm12 replied:
> If I use @ or * it updates all DDNS IPs for my domains on Dynu, * doesn't update subdomains of a concrete domain, so *.domain.com updates domain.com but not subdomain.domain.com

I'm a bit confused here.
Correct me if I'm wrong:
- `@` for domain.com updates only `domain.com` but not `sub.domain.com
- `*` for domain.com updates `domain.com`, `domain2.com` but not `sub.domain.com`

msxdan replied:
Well, `@` does nothing really, if you don't set the `hostname` or if it's not a valid hostname for your user it will update all records for your domains that don't have a group assigned.

`*` doesn't acts like a wildcard, you can use it in the hostname but it will update only the `domain.com` and not `sub.domain.com`

qdm12 commented on internal/settings/providers/dynu/provider.go:
```diff
@@ -0,0 +1,155 @@
+	case strings.Contains(s, "nochg"): // Updated but not changed
+		return ip, nil
+	default:
+		return nil, fmt.Errorf("%w: %s", errors.ErrUnknownResponse, s)
```
```suggestion
		return nil, fmt.Errorf("%w: %s", errors.ErrUnknownResponse, utils.ToSingleLine(s))
```

qdm12 reviewed (approved):
Nice, thanks for the contribution!!! Especially on new year, impressive :tada: !

Just a few nit-picking comments

msxdan commented:
I finally added support for most options on Dynu, feel free to review and give any advices about code

![image](https://user-images.githubusercontent.com/32611659/148305054-68dddcce-952b-4368-be41-64e30d5888f0.png)

Example usage: (Comments are for reference)

```
{
    "settings": [
        { // Subdomain update [any.host.dynu.net]
            "provider": "dynu",
            "domain": "dynu.net",
            "host": "host",
            "alias": "any",
            "username" : "user",
            "password": "63a9f0ea7bb98050796b649e85481845",
            "ip_version": "ipv4",
            "provider_ip": true
        },
        { // Subdomain update [www.host.dynu.net]
            "provider": "dynu",
            "domain": "dynu.net",
            "host": "host",
            "alias": "www",
            "username" : "user",
            "password": "63a9f0ea7bb98050796b649e85481845",
            "ip_version": "ipv4",
            "provider_ip": true
        },
        { // domain update [host.dynu.net]
            "provider": "dynu",
            "domain": "dynu.net",
            "host": "host",
            "alias": "",
            "username" : "user",
            "password": "63a9f0ea7bb98050796b649e85481845",
            "ip_version": "ipv4",
            "provider_ip": true
        },
        { // Group update [host.dynu.net | group=work]
            "provider": "dynu",
            "domain": "dynu.net",
            "host": "host",
            "alias": "",
            "location": "work",
            "username" : "user",
            "password": "63a9f0ea7bb98050796b649e85481845",
            "ip_version": "ipv4",
            "provider_ip": true
        },
        { // Update all domains you have in dynu that doesn't belongs to any group
            "provider": "dynu",
            "domain": "host.dynu.net",
            "host": "@",
            "username" : "user",
            "password": "63a9f0ea7bb98050796b649e85481845",
            "ip_version": "ipv4",
            "provider_ip": true
        }
    ]
}
```

qdm12 commented on internal/settings/providers/dynu/provider.go:
```diff
@@ -53,25 +59,30 @@ func New(data json.RawMessage, domain, host string,
 func (p *provider) Host() string {
+	if p.alias != "" {
+		return p.alias + "." + p.host
+	}
```
Are you sure this is fine?

Should we return alias.host instead of just alias? 🤔

msxdan commented:
I removed the mess with extra params and simplified the usage, adding only an extra parameter for the group.

![image](https://user-images.githubusercontent.com/32611659/148403079-08d9562b-588c-4913-ae6e-f5e2c740ae4d.png)

```
{
    "settings": [
        {
            "provider": "dynu",
            "domain": "dom.dynu.net",
            "host": "@",
            "username" : "user",
            "password": "<md5hash>",
            "ip_version": "ipv4",
            "provider_ip": false,
            "ip": "10.1.0.1"
        },
        {
            "provider": "dynu",
            "domain": "dom.dynu.net",
            "host": "www", // Subdomain
            "username" : "user",
            "password": "<md5hash>",
            "ip_version": "ipv4",
            "provider_ip": false,
            "ip": "10.1.0.2"
        },
        {
            "provider": "dynu",
            "domain": "dom.dynu.net",
            "host": "any", // Subdomain
            "username" : "user",
            "password": "<md5hash>",
            "ip_version": "ipv4",
            "provider_ip": false,
            "ip": "10.1.0.5"
        },
        {
            "provider": "dynu",
            "domain": "dom2.dynu.net",
            "host": "sub",
            "group": "work", // It will check the subdomain IP and update the group
            "username" : "user",
            "password": "<md5hash>",
            "ip_version": "ipv4",
            "provider_ip": false,
            "ip": "10.1.0.4"
        }
    ]
}
```

msxdan commented:
Any news about these changes? I'm currently using it and it's working fine, we could merge it into master if you agree

qdm12 commented on internal/settings/providers/dynu/provider.go:
```diff
@@ -0,0 +1,166 @@
+		values.Set("alias", p.host)
+		values.Set("hostname", p.domain)
+	} else {
+		values.Set("hostname", utils.BuildURLQueryHostname(p.host, p.domain))
```
Isn't this just 

```suggestion
		values.Set("hostname", p.domain)
```

Since `p.host` is `@`?

qdm12 commented on internal/settings/providers/dynu/provider.go:
```diff
@@ -0,0 +1,166 @@
+	if err := json.Unmarshal(data, &extraSettings); err != nil {
+		return nil, err
+	}
+	p = &provider{
```
Please default the host to `"@"` and remove the error `ErrEmptyHost`

```suggestion

	if host == "" {
		host = "@" // default
	}
	
	p = &provider{
```

qdm12 commented on internal/settings/providers/dynu/provider.go:
```diff
@@ -0,0 +1,166 @@
+		return errors.ErrEmptyUsername
+	case p.password == "":
+		return errors.ErrEmptyPassword
+	case p.host == "":
+		return errors.ErrEmptyHost
```
```suggestion
```

qdm12 commented on internal/settings/providers/dynu/provider.go:
```diff
@@ -0,0 +1,166 @@
+
+	request, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
+	if err != nil {
+		return nil, err
```
```suggestion
		return nil, fmt.Errorf("%w: %s", errors.ErrBadRequest, err)
```

qdm12 commented on internal/settings/providers/dynu/provider.go:
```diff
@@ -0,0 +1,166 @@
+
+	response, err := client.Do(request)
+	if err != nil {
+		return nil, err
```
```suggestion
		return nil, fmt.Errorf("%w: %s", errors.ErrUnsuccessfulResponse, err)
```

qdm12 reviewed (commented):
Sorry for the delay this got lost in my emails 😢 
Just a few comments and let's merge this ASAP 😉 
Thanks again for your work!

C-SELLERS commented:
Just built and am using an image of this branch for Dynu as well, everything seems to working fine. Thanks guys 😁

msxdan commented:
Done!

qdm12 commented:
Awesome thanks for your work and dedication @msxdan
